runtime/backtrace_nat.c

```diff
@@ -121,12 +121,23 @@ intnat caml_current_callstack_size(intnat max_frames) {
   return trace_size;
 }
 
-void caml_current_callstack_write(value trace) {
+static debuginfo debuginfo_extract(frame_descr* d, unsigned alloc_idx);
+void caml_current_callstack_write(value trace, unsigned alloc_idx)
+{
   uintnat pc = Caml_state->last_return_address;
   char * sp = Caml_state->bottom_of_stack;
-  intnat trace_pos, trace_size = Wosize_val(trace);
+  intnat trace_pos = 0, trace_size = Wosize_val(trace);
+
+  if (alloc_idx > 0 && trace_size > 0) {
+    frame_descr * descr = caml_next_frame_descriptor(&pc, &sp);
+    debuginfo info = debuginfo_extract(descr, alloc_idx);
+    /* Mark this debuginfo as already extracted */
+    backtrace_slot slot = (backtrace_slot)((uintnat)info + 2);
```

Comment:
I would find this logic (and `caml_debuginfo_extract`) clearer with macros to give names to the bit-fiddling operations.

Some proposals would be for example:
```
Val_direct_debuginfo : debuginfo -> val
Direct_debuginfo_of_val : val -> debuginfo
Is_direct_debuginfo(_or_backtrace_slot) : val -> bool

Slot_of_direct_debuginfo : debuginfo -> backtrace_slot
Slot_is_direct_debuginfo : backtrace_slot -> bool
Extract_direct_debuginfo : backtrace_slot -> debuginfo
```

Reply:
Good point, changed. (I picked different names to match the runtime's `To_from` convention for conversion macros)

Reply:
Thanks! Your names are good.